### PR TITLE
feat(datasets): add AddDataSetAccessPermission to DataSetService

### PR DESCRIPTION
**Pull request by DaveMadsen**
Issue #, if available:

Description of changes:
Add the ability to add an access permission for a DataSet to the DataSetsService

Checklist:

* [ X] Have you successfully deployed to an AWS account with your changes?
* [X ] Have you written new tests for your core changes, as applicable?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: workbench-core/datasets/src/dataSetService.test.ts
Comment (aws-tyler) [inline workbench-core/datasets/src/dataSetService.test.ts:145]: is it possible to move this logic to specific `describe` sections for each test context so its clear which mock results belong to a specific test?

File: workbench-core/datasets/src/dataSetService.test.ts
Comment (aws-tyler) [inline workbench-core/datasets/src/dataSetService.test.ts:513]: I think a `rejects.toThrow` check will only compare the message that gets passed in, not the HTTP error. If you change this to another Boom error does it still pass? I think there is a `checkHTTPError()` method that gets used to verify when an HTTP error gets thrown that you may be able to use here.

File: workbench-core/datasets/src/dataSetService.ts
Comment (aws-tyler) [inline workbench-core/datasets/src/dataSetService.ts:78]: Why are we hard-coding this to be a 400? Could it ever be another 4xx or 5xx error?

File: workbench-core/datasets/src/dataSetService.ts
Comment (aws-tyler) [inline workbench-core/datasets/src/dataSetService.ts:74]: Should this be a 201 since we're creating a resource?

File: workbench-core/datasets/src/dataSetService.ts
Comment (aws-tyler) [inline workbench-core/datasets/src/dataSetService.ts:56]: is `-` a placeholder value?

File: workbench-core/datasets/src/dataSetService.ts
Comment (aws-tyler) [inline workbench-core/datasets/src/dataSetService.ts:174]: Is `dataSetId` a requestBody value even if the param is coming from the url and not the body?

File: workbench-core/example/infrastructure/integration-tests/support/resources/datasets/dataset.ts
Comment (aws-tyler) [inline workbench-core/example/infrastructure/integration-tests/support/resources/datasets/dataset.ts:57]: should this be set via a switch that throws an error if an invalid value gets passed into to the method?  since `requestBody` isn't typed we can't guarantee we're getting valid data.

File: workbench-core/example/infrastructure/integration-tests/support/resources/datasets/dataset.ts
Comment (aws-tyler) [inline workbench-core/example/infrastructure/integration-tests/support/resources/datasets/dataset.ts:61]: same as above

File: workbench-core/example/infrastructure/integration-tests/tests/datasets/accessPermissions.test.ts
Comment (aws-tyler) [inline workbench-core/example/infrastructure/integration-tests/tests/datasets/accessPermissions.test.ts:41]: we should keep these immutable for safety

File: workbench-core/example/infrastructure/integration-tests/tests/datasets/accessPermissions.test.ts
Comment (aws-tyler) [inline workbench-core/example/infrastructure/integration-tests/tests/datasets/accessPermissions.test.ts:64]: This confuses me a little bit: how can you have `read-only` AND `read-write` permissions on the same dataset? They're conflicting permissions, right?

File: workbench-core/example/infrastructure/integration-tests/tests/datasets/accessPermissions.test.ts
Comment (aws-tyler) [inline workbench-core/example/infrastructure/integration-tests/tests/datasets/accessPermissions.test.ts:48]: i see a lot of copy and pasted strings: do we have constants for the domains specific strings for AuthZ like `GROUP`, `read-only`, `read-write`, etc?

File: workbench-core/example/infrastructure/integration-tests/tests/datasets/accessPermissions.test.ts
Comment (aws-tyler) [inline workbench-core/example/infrastructure/integration-tests/tests/datasets/accessPermissions.test.ts:67]: do you still need this?

File: workbench-core/example/infrastructure/integration-tests/tests/datasets/accessPermissions.test.ts
Comment (aws-tyler) [inline workbench-core/example/infrastructure/integration-tests/tests/datasets/accessPermissions.test.ts:44]: why do we need the `as` here?

File: workbench-core/datasets/src/dataSetService.test.ts
Comment (DaveMadsen) [reply workbench-core/datasets/src/dataSetService.test.ts:145]: fun fact, I had things all separated out long long ago and was asked to push it all up front. I don't think I have time to unwind that today.

File: workbench-core/datasets/src/dataSetService.test.ts
Comment (DaveMadsen) [reply workbench-core/datasets/src/dataSetService.test.ts:513]: interesting, and fix coming in next push.

File: workbench-core/datasets/src/dataSetService.ts
Comment (DaveMadsen) [reply workbench-core/datasets/src/dataSetService.ts:56]: In this case, as we will need those for creating the access permissions.
There is another spot where we aren't and for now, won't be passed the authenticated user however someone made `actor` a required field.

Comment (DaveMadsen) [reply workbench-core/datasets/src/dataSetService.ts:56]: oh wait, it's optional! I'll just remove it until I'm ready for it.

File: workbench-core/datasets/src/dataSetService.ts
Comment (DaveMadsen) [reply workbench-core/datasets/src/dataSetService.ts:74]: I've asked that on other PRs. The response I've received is along the lines of "well, we don't determine a status code when we throw the error, and we leave it as an exercise to the consumer to choose a status code when they return a result. 200 and 400 are just supposed to represent generic success/failure."

I don't like it either. Working from the "we don't do status codes here" argument, I've removed the status code in the next push.

File: workbench-core/datasets/src/dataSetService.ts
Comment (DaveMadsen) [reply workbench-core/datasets/src/dataSetService.ts:78]: see previous comment. Expect this to go away.

File: workbench-core/datasets/src/dataSetService.ts
Comment (DaveMadsen) [reply workbench-core/datasets/src/dataSetService.ts:174]: changing to params

File: workbench-core/example/infrastructure/integration-tests/support/resources/datasets/dataset.ts
Comment (DaveMadsen) [reply workbench-core/example/infrastructure/integration-tests/support/resources/datasets/dataset.ts:61]: validation added.

File: workbench-core/example/infrastructure/integration-tests/tests/datasets/accessPermissions.test.ts
Comment (DaveMadsen) [reply workbench-core/example/infrastructure/integration-tests/tests/datasets/accessPermissions.test.ts:41]: you mean the responses? In test code? I'll take a look.

File: workbench-core/example/infrastructure/integration-tests/tests/datasets/accessPermissions.test.ts
Comment (DaveMadsen) [reply workbench-core/example/infrastructure/integration-tests/tests/datasets/accessPermissions.test.ts:44]: `children` if of type `Map<string,Resource>`. `Dataset` (in this case) extends `Resource`.

File: workbench-core/example/infrastructure/integration-tests/tests/datasets/accessPermissions.test.ts
Comment (DaveMadsen) [reply workbench-core/example/infrastructure/integration-tests/tests/datasets/accessPermissions.test.ts:48]: I will look into this. I can use constants if we define them in `base` or `datasets`. I prefer not to use them if they are defined in `authz` as that will encourage making the package coupling stronger. I would eventually like to refactor our code to decouple more.

File: workbench-core/example/infrastructure/integration-tests/tests/datasets/accessPermissions.test.ts
Comment (DaveMadsen) [reply workbench-core/example/infrastructure/integration-tests/tests/datasets/accessPermissions.test.ts:64]: I can change this behavior. 

read-only from an authz perspective is simply `READ`.
read-write is `READ` and `UPDATE` (and if you need them, probably `CREATE`/`DELETE`, but not needed at this juncture for our purposes).

I can create a private method to manage the conversion from READ to read-only and READ/UPDATE to read-write. I originally just used `UPDATE` to capture read-write, but somebody else pushed back on that.

File: workbench-core/example/infrastructure/integration-tests/tests/datasets/accessPermissions.test.ts
Comment (DaveMadsen) [reply workbench-core/example/infrastructure/integration-tests/tests/datasets/accessPermissions.test.ts]: nope

**Review by aws-tyler — DISMISSED** [review]

File: workbench-core/example/infrastructure/integration-tests/tests/datasets/accessPermissions.test.ts
Comment (DaveMadsen) [reply workbench-core/example/infrastructure/integration-tests/tests/datasets/accessPermissions.test.ts]: okay. finally! I ended up doing some future proofing and separated out converting the dynamic auth output into datasets obejcts into a private method which needed a bunch of unit testing of its own. It's a bit more robust than what I had too.

But now, the READ permission on read-write access will be ignored and UPDATE will supersede with the read-write response.

**Review by aws-tyler — APPROVED** [review]

File: workbench-core/datasets/src/dataSetService.ts
Comment (htztomic) [inline workbench-core/datasets/src/dataSetService.ts:73]: metadata needs to be updated with statusCode

File: workbench-core/datasets/src/dataSetService.ts
Comment (htztomic) [inline workbench-core/datasets/src/dataSetService.ts:55]: will also need to include an actor

File: workbench-core/datasets/src/dataSetService.ts
Comment (htztomic) [inline workbench-core/datasets/src/dataSetService.ts:166]: could this be more specific?

File: workbench-core/datasets/src/dataSetService.ts
Comment (htztomic) [inline workbench-core/datasets/src/dataSetService.ts:374]: status code missing here

File: workbench-core/datasets/src/models/dataSetPermission.ts
Comment (htztomic) [inline workbench-core/datasets/src/models/dataSetPermission.ts:12]: IdentityPermission is defined with this being identityId

File: workbench-core/datasets/src/models/dataSetPermission.ts
Comment (htztomic) [inline workbench-core/datasets/src/models/dataSetPermission.ts:14]:
```suggestion
  identityType: z.union([z.literal('GROUP'), z.literal('USER')]);,
```

File: workbench-core/datasets/src/dataSetService.ts
Comment (DaveMadsen) [reply workbench-core/datasets/src/dataSetService.ts:73]: If it _needs_ to be here, why is it optional on the metadata object?
I don't know what status code will be returned yet. I am not doing a res.send() here.
200/400 are not necessarily accurate and will make the logs confusing.
what do?

File: workbench-core/datasets/src/dataSetService.ts
Comment (DaveMadsen) [reply workbench-core/datasets/src/dataSetService.ts:55]: Not on this PR, but will come in when this method is updated to add an access permission.

File: workbench-core/datasets/src/dataSetService.ts
Comment (DaveMadsen) [reply workbench-core/datasets/src/dataSetService.ts:166]: Not without a late game breaking change to a method we did not claim to be updating.

File: workbench-core/datasets/src/dataSetService.ts
Comment (DaveMadsen) [reply workbench-core/datasets/src/dataSetService.ts:374]: see previous comment about status codes being optional and confusing.

File: workbench-core/datasets/src/dataSetService.ts
Comment (htztomic) [reply workbench-core/datasets/src/dataSetService.ts:73]: So it is optional as an input, but the auditService will verify if the audit entry is complete before writing it. This is to ensure all audits are following best practices and creating a valid audit trail. By default, auditService will require actor, source, statusCode, and action

File: workbench-core/datasets/src/models/dataSetPermission.ts
Comment (DaveMadsen) [reply workbench-core/datasets/src/models/dataSetPermission.ts:14]: Intentionally avoided at this level.

I leave it up to the implementation of the authorization plugin to choose the identity types appropriate for the authz backend and the specific use case. Those are correct for our reference implementation, but defining that union in at this interface level limits future implementations.

File: workbench-core/datasets/src/dataSetService.ts
Comment (DaveMadsen) [reply workbench-core/datasets/src/dataSetService.ts:73]: so you're saying that I will need to make a breaking change on every public API to add auditing?

File: workbench-core/datasets/src/wbcDataSetsAuthorizationPlugin.ts
Comment (htztomic) [inline workbench-core/datasets/src/wbcDataSetsAuthorizationPlugin.ts:77]: when we expect this error, are we expecting users to delete those permissions?

File: workbench-core/datasets/src/dataSetService.ts
Comment (htztomic) [reply workbench-core/datasets/src/dataSetService.ts:73]: not necessarily, actor is dependent on what you determine is sufficient enough to do an audit trial on. If we require the user that is requesting this information to be the actor then yes, we will need that information.

File: workbench-core/datasets/src/wbcDataSetsAuthorizationPlugin.ts
Comment (DaveMadsen) [reply workbench-core/datasets/src/wbcDataSetsAuthorizationPlugin.ts:74]: we are not expecting this error.

**Review by htztomic — APPROVED** [review]
LGTM

**Review by aws-tyler — APPROVED** [review]

**Pull request merged**
